head/src/main/kotlin/io/qalipsis/core/head/report/DataSeriesServiceImpl.kt

```diff
@@ -196,4 +204,48 @@ internal class DataSeriesServiceImpl(
         )
     }
 
+    override suspend fun refresh() {
+        backgroundScope.launch {
+            var pageable = Pageable.from(0, SIZE)
+            var pagedDataSeries = dataSeriesRepository.findDataSeries(pageable)
+            handleQueryUpdate(pagedDataSeries)
+            while (pagedDataSeries.pageNumber < pagedDataSeries.totalPages - 1) {
+                pageable = pagedDataSeries.nextPageable()
+                pagedDataSeries = dataSeriesRepository.findDataSeries(pageable)
+                handleQueryUpdate(pagedDataSeries)
+            }
+        }
+    }
+
+    /**
+     * Handle the update of each query for a given batch of data series entity.
+     */
+    private suspend fun handleQueryUpdate(pagedDataSeries: Page<DataSeriesEntity>) {
+        var countOfUpdatedQueries = pagedDataSeries.numberOfElements
+        try {
+            logger.trace { "Updating the queries of ${pagedDataSeries.numberOfElements} retrieved data series" }
+            val dataSeriesEntities = pagedDataSeries.content
+            dataSeriesEntities.forEach { dataSeriesEntity ->
+                val tenant = tenantRepository.findReferenceById(dataSeriesEntity.tenantId)
+                dataSeriesEntity.query = dataProvider.createQuery(
+                    tenant, dataSeriesEntity.dataType, QueryDescription(
+                        filters = convertFiltersEntity(dataSeriesEntity.filters, dataSeriesEntity.valueName),
+                        fieldName = dataSeriesEntity.fieldName,
+                        aggregationOperation = dataSeriesEntity.aggregationOperation,
+                        timeframeUnit = dataSeriesEntity.timeframeUnitAsDuration
+                    )
+                ).takeUnless(String::isNullOrBlank)
+            }
+            dataSeriesRepository.saveAll(dataSeriesEntities).count()
+            logger.info { "Successfully updated $countOfUpdatedQueries queries of the data series" }
+        } catch (ex: Exception) {
+            countOfUpdatedQueries -= 1
```

Comment:
Why -1? How do you know here how many failed? Aren't they actually `countOfUpdatedQueries` failing?

Reply:
The initial value of  `countOfUpdatedQueries`  is `pagedDataSeries.numberOfElements`. So for any that fails I subtract 1 from the total. At the end, what we have as countOfUpdatedQueries is the number that was successfully updated without fail.
Now that I think about it though it might not be the best descriptive name

Reply:
bit there is no look in that call, and the save is for the batch...